### PR TITLE
Keep assembly functions in separate sections

### PR DESCRIPTION
**Pull request by jenswi-linaro**
To get a more fine grained selection of which area (init, paged,
unpaged) an assembly function is assigned do the equivalent of
-ffunction-sections but in assembly.

Some functions has to be in specific places in the binary for a
successful boot, link script is updated accordingly.

Tested-by: Jens Wiklander <jens.wiklander@linaro.org> (QEMU v7/v8)
Signed-off-by: Jens Wiklander <jens.wiklander@linaro.org>

### CODE REVIEW

File: core/arch/arm/kernel/generic_entry_a64.S [hunk 1]
```diff
@@ -234,6 +233,7 @@ END_FUNC unhandled_cpu
 	  .endif
 	.endm
 
+	.section .text.reset_vect_table
```
Comment (jforissier) [inline core/arch/arm/kernel/generic_entry_a64.S]: Not needed?

File: core/arch/arm/kernel/thread_a32.S [hunk 2]
```diff
@@ -817,6 +819,7 @@ UNWIND(	.cantunwind)
 UNWIND(	.fnend)
 END_FUNC thread_svc_handler
 
+	.section .text.thread_vect_table
```
Comment (jforissier) [inline core/arch/arm/kernel/thread_a32.S]: Not needed?

File: core/arch/arm/kernel/thread_a64.S [hunk 3]
```diff
@@ -345,6 +346,7 @@ END_FUNC thread_unwind_user_mode
 	.endm
 
 
+	.section .text.thread_vect_table
```
Comment (jforissier) [inline core/arch/arm/kernel/thread_a64.S]: Not needed?

File: core/arch/arm/sm/pm_a32.S [hunk 4]
```diff
@@ -147,6 +147,20 @@ UNWIND(	.fnend)
...
+ */
+.align 2
+.extern thread_core_local
+_suspend_sp:
+	.long	thread_core_local - .
+.extern get_core_pos
+_core_pos:
+	.long	get_core_pos - .
+
+/*
```
Comment (jforissier) [inline core/arch/arm/sm/pm_a32.S]: Why move this block here? Just cosmetic?

File: core/arch/arm/sm/sm_a32.S [hunk 5]
```diff
@@ -244,6 +244,7 @@ UNWIND(	.cantunwind)
 UNWIND(	.fnend)
 END_FUNC sm_fiq_entry
 
+	.section .text.sm_vect_table
```
Comment (jforissier) [inline core/arch/arm/sm/sm_a32.S]: Not needed?

File: core/arch/arm/sm/pm_a32.S (hunk 4 above)
Comment (jenswi-linaro) [reply core/arch/arm/sm/pm_a32.S]: It's to avoid an assembler error, apparantly the `adr` instruction can only be performed on symbols in the same section.

**Comment by jenswi-linaro** [comment]
"core: arm32: replace _start with reset() function" is needed also for this to have full effect.

**Comment by jenswi-linaro** [comment]
Rebased

File: core/arch/arm/kernel/thread_a32.S [hunk 6]
```diff
@@ -210,7 +211,7 @@ UNWIND(	.cantunwind)
 	msr	cpsr, r1
 	bx	lr
 UNWIND(	.fnend)
-END_FUNC thread_set_abt_sp
+END_FUNC thread_set_und_sp
```
Comment (jforissier) [inline core/arch/arm/kernel/thread_a32.S]: Nit: unrelated fix, should be a separate commit

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 7]
```diff
@@ -529,12 +528,15 @@ UNWIND(	.cantunwind)
...
 END_FUNC cpu_on_handler
+KEEP_PAGER cpu_on_handler
 
 #else /* defined(CFG_WITH_ARM_TRUSTED_FW) */
 
 LOCAL_FUNC reset_secondary , :
 UNWIND(	.fnstart)
 UNWIND(	.cantunwind)
+	ldr	r0, =reset_vect_table
+	write_vbar r0
```
Comment (jforissier) [inline core/arch/arm/kernel/generic_entry_a32.S]: Isn't this an unrelated fix? (would deserve its own commit)

**Comment by jenswi-linaro** [comment]
Split the commits

**Comment by jforissier** [comment]
Thanks for splitting the commits.
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jenswi-linaro** [comment]
Tag applied
